ESQL: Implement LOOKUP, an "inline" enrich

nik9000 opened the pull request:
This adds support for `LOOKUP`, a command that implements a sort of
inline `ENRICH`, using data that is passed in the request:

```
$ curl -uelastic:password -HContent-Type:application/json -XPOST \
    'localhost:9200/_query?error_trace&pretty&format=txt' \
-d'{
    "query": "ROW a=1::LONG | LOOKUP t ON a",
    "tables": {
        "t": {
            "a:long":     [    1,     4,     2],
            "v1:integer": [   10,    11,    12],
            "v2:keyword": ["cat", "dog", "wow"]
        }
    },
    "version": "2024.04.01"
}'
      v1       |      v2       |       a       
---------------+---------------+---------------
10             |cat            |1              
```



This required these PRs:
* #107624
* #107634
* #107701
* #107762
* #107923
* #107894
* #107982
* #108012
* #108020
* #108169
* #108191
* #108334
* #108482
* #108696
* #109040
* #109045

Closes #107306

nik9000 committed:
WIP

nik9000 committed:
WIP

nik9000 committed:
WIP2

nik9000 committed:
Merge branch 'main' into lookup_real___

nik9000 committed:
Links

nik9000 committed:
Update docs/changelog/107987.yaml

elasticsearchmachine commented:
Hi @nik9000, I've created a changelog YAML for you.

nik9000 committed:
Rename

nik9000 commented:
I've opened this as draft because I've not added any real tests. And because stuff like:
```
curl -uelastic:password -HContent-Type:application/json -XPOST 'localhost:9200/_query?error_trace&pretty' -d'{
    "query": "ROW a=1::LONG | LOOKUP t1 ON a | LOOKUP t2 ON v1",
    "tables": {
        "t1": {
            "a:long":     [1, 4, 2],
            "v1:long":    [5, 8, 9]
        },
        "t2": {
            "v1:long":    [    5,     8,     9],
            "v2:keyword": ["cat", "dog", "wow"]
        }
    },
    "version": "2024.04.01"
}'
```
fails with array index out of bounds exceptions that don't make sense to me yet.

nik9000 committed:
doc

nik9000 committed:
Merge remote-tracking branch 'nik9000/lookup_real___' into lookup_real___

nik9000 committed:
More docs

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -312,10 +318,13 @@ private static NamedExpression createEnrichFieldExpression(
     }
 
-    private static class ResolveRefs extends BaseAnalyzerRule {
+    private static class ResolveRefs extends ParameterizedAnalyzerRule<LogicalPlan, AnalyzerContext> {
```
Is this ok? I need the context.

costin replied:
Better to use the approach for Enrich - create a separate rule that gets the analyzer context and partially/fully resolves the lookup since all information for it is already available.
If there are extra things, complete the resolution inside ResolveRefs (without having to pass the context)  - essentially the first rule does a first pass, letting ResolveRefs to fail the resolution if necessary.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -407,6 +420,52 @@ private LogicalPlan resolveAggregate(Aggregate a, List<Attribute> childrenOutput
+            }
+            table = new TreeMap<>(table);
+            // TODO what's the right type for this thing?!
+            Attribute resolvedTableName = new MetadataAttribute(p.tableName().source(), tableName, KEYWORD, false);
```
I have no idea what the appropriate `Attribute` subclass is here - we don't really use it for anything after this other than, like, `toString` and making sure all of the attributes are `resolved`.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -407,6 +420,52 @@ private LogicalPlan resolveAggregate(Aggregate a, List<Attribute> childrenOutput
+                }
+
+                matchFields.add(m);
+                matchValues.add(new TableColumnAttribute(m.source(), m.name(), c.type(), null, c.values()));
```
I added a new `Attribute` subclass that adds the values to it.

nik9000 replied:
I've removed this one.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/OptimizerRules.java:
```diff
@@ -105,6 +106,10 @@ protected AttributeSet references(LogicalPlan plan) {
             }
+            if (plan instanceof Lookup lookup) {
+                // NOCOMMIT what in the world am I doing?
+                return Expressions.references(lookup.matchFields());
```
I have no idea why I had to do this. It feels wrong.

costin replied:
Likely because the match fields are not found in the output - this is indeed a red flag that we should look into.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/ExpressionBuilder.java:
```diff
@@ -223,6 +223,11 @@ public UnresolvedAttribute visitQualifiedName(EsqlBaseParser.QualifiedNameContex
+    @Override
+    public List<Attribute> visitQualifiedNames(EsqlBaseParser.QualifiedNamesContext ctx) {
+        return visitList(this, ctx.qualifiedName(), Attribute.class);
+    }
```
Do we need this? Is there some other rule I should be using instead?

costin replied:
You shouldn't - the default implementation should automatically do the right thing. I assume you tried it though and it failed - what was the error?

nik9000 replied:
They come back the wrong type. I'm wondering if there's a different `visit` I should be using - it feels strange that I'm the first one to want names in the shape that I want them. I'll bet there's something else similar and I just didn't see it.

nik9000 committed:
Explain more

nik9000 committed:
Update docs/changelog/107987.yaml

elasticsearchmachine commented:
Hi @nik9000, I've updated the changelog YAML for you.

nik9000 committed:
Fix a silly

nik9000 committed:
Merge remote-tracking branch 'nik9000/lookup_real___' into lookup_real___

nik9000 commented:
So I ran a little test:
```
curl -uelastic:password -XDELETE localhost:9200/test
for a in {0..99}; do
   rm -f /tmp/evil
   for b in {0..999}; do
      echo '{"index": {}}' >> /tmp/evil
      echo '{"a": 1}' >> /tmp/evil
      echo '{"index": {}}' >> /tmp/evil
      echo '{"a": 4}' >> /tmp/evil
      echo '{"index": {}}' >> /tmp/evil
      echo '{"a": 2}' >> /tmp/evil
   done
   echo >> /tmp/evil
   echo -n "$a:  "
   curl -s -HContent-Type:application/json -uelastic:password -XPOST localhost:9200/test/_bulk?pretty --data-binary @/tmp/evil | grep \"errors\"
done

curl -HContent-Type:application/json -uelastic:password -XPOST localhost:9200/test/_forcemerge?max_num_segments=1
curl -HContent-Type:application/json -uelastic:password -XPOST localhost:9200/test/_refresh
```

Then I ran these two a bunch:
```
curl -uelastic:password -HContent-Type:application/json -XPOST \
    'localhost:9200/_query?error_trace&pretty' \
-d'{
    "query": "FROM test | EVAL v1=a+1 | STATS COUNT(v1)",
    "profile": true,
    "version": "2024.04.01"
}'
curl -uelastic:password -HContent-Type:application/json -XPOST \
    'localhost:9200/_query?error_trace&pretty' \
-d'{
    "query": "FROM test | LOOKUP t ON a | STATS COUNT(v1)",
    "profile": true,
    "tables": {
        "t": {
            "a:long":     [    1,     4,     2],
            "v1:int":     [   10,    11,    12]
        }
    },
    "version": "2024.04.01"
}'
```

That's 300,000 documents - not many, but small enough to test fast on my laptop. From the profile `EVAL v1=a + 1` took 1ms or about 3 nanoseconds per value. The actual addition is likely less than .3ns per operation but I guess the block accounting gets us slower. `LOOKUP t ON a` took 10ms or about 33ns per value. That's not scientific, but it's believable. About 28ns of those 33ns is hashing - which isn't a huge surprise - we know we're using our slowest hash implementation. We have faster hash implementations. We just have to flip a few things around to plug them in properly.

nik9000 commented:
My last push seems to have fixed double-layered lookup:
```
curl -uelastic:password -HContent-Type:application/json -XPOST 'localhost:9200/_query?error_trace&pretty' -d'{
    "query": "ROW a=1::LONG | LOOKUP t1 ON a | LOOKUP t2 ON v1",
    "tables": {
        "t1": {
            "a:long":     [1, 4, 2],
            "v1:long":    [5, 8, 9]
        },
        "t2": {
            "v1:long":    [    5,     8,     9],
            "v2:keyword": ["cat", "dog", "wow"]
        }
    },
    "version": "2024.04.01"
}'
```

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/ColumnLoadOperator.java:
```diff
@@ -0,0 +1,67 @@
+ * {@link Block#lookup Looks up} values from a provided {@link Block} and
+ * mergeds them into each {@link Page}.
+ */
+public class ColumnLoadOperator extends AbstractPageMappingToIteratorOperator {
```
I've yanked this into it's own PR: https://github.com/elastic/elasticsearch/pull/108020

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/HashLookupOperator.java:
```diff
@@ -122,8 +123,13 @@ public void add(int positionOffset, IntVector groupIds) {
+            return appendBlocks(page, hash.lookup(mapped, BlockFactory.DEFAULT_MAX_BLOCK_PRIMITIVE_ARRAY_SIZE));
+        } finally {
+            mapped.releaseBlocks();
+        }
```
NOCOMMIT: Add a unit test that'll fail without this change.

nik9000 replied:
#108191

nik9000 commented:
This needs a bunch more tests and docs.

This is a feature that'll likely stay in experimental for a few versions regardless of how difficult it is for us to settle the external facing stuff.

nik9000 committed:
Merge branch 'main' into lookup_real___

nik9000 committed:
Tests

nik9000 committed:
Fix more tests

nik9000 commented:
I'm getting failures in `NodeSubclassTests`. It'd be nice to have a little help on that one. I'm not sure what it's doing.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/ParseTables.java:
```diff
@@ -0,0 +1,244 @@
+                        return columns;
+                    }
+                    case FIELD_NAME -> {
+                        // NOCOMMIT just infer type from the json
```
I'm not entirely sure we want this. I mean, maybe. But stuff like `DATE`s would be nice to be able to parse one time.

costin replied:
This open the discussion about what is the data format allowed on the wire (is CSV, TSV, arrow, etc...)?
IF we end up going this route, I expect we'll have a qualifier to indicate the resource type (CSV, etc..).

nik9000 committed:
Don't revert this

costin commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseLexer.g4:
```diff
@@ -343,6 +344,46 @@ ENRICH_FIELD_WS
+
+LOOKUP_WS
+    : WS -> channel(HIDDEN)
+    ;
```
LGTM

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/TableColumnAttribute.java:
```diff
@@ -0,0 +1,69 @@
+/**
+ * An {@link Attribute} that contains a {@link Block} of values.
+ */
+public class TableColumnAttribute extends TypedAttribute {
```
Better to wrap this in a LocalRelation and have the blocks (for each column) supplied through LocalSupplier.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Lookup.java:
```diff
@@ -0,0 +1,129 @@
+ *     up the table from the analysis context.
+ * </p>
+ */
+public class Lookup extends UnaryPlan {
+    private final Attribute tableName;
+    private final List<Attribute> matchFields;
+    private final List<TableColumnAttribute> matchValues;
+    private final List<TableColumnAttribute> mergeValues;
+    private List<Attribute> output;
+
+    public Lookup(
+        Source source,
+        LogicalPlan child,
+        Attribute tableName,
+        List<Attribute> matchFields,
+        @Nullable List<TableColumnAttribute> matchValues,
+        @Nullable List<TableColumnAttribute> mergeValues
+    ) {
+        super(source, child);
+        this.tableName = tableName;
+        this.matchFields = matchFields;
+        this.matchValues = matchValues;
+        this.mergeValues = mergeValues;
+    }
```
Like Enrich this class tries to do two things:
- define the stiching/join
- act as a source (without being one) for the broadcast side.
This works but same limitations - let's discuss this offline.

costin reviewed (commented):
Left a round of comments.

nik9000 committed:
Remove lexer bits

nik9000 committed:
Merge branch 'main' into lookup_real___

nik9000 committed:
integration test

nik9000 committed:
Add a test

it doesn't pass

nik9000 commented:
Costin's going to juggle the the language parts of this a fair bit.

costin committed:
Model Lookup as a join with a local relation

Model the configuration table as a LocalRelation
Analyzer resolves the matching field on both child and local relation
 that is nested (to preserve the definition of Lookup which is a
 UnaryPlan)
LogicalPlanOptimizer substitutes the lookup with a join and promotes the
 local relation into the tree.
Refactor some common parts of the grammar to make them more consistent

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/EsqlTestUtils.java:
```diff
@@ -119,10 +118,53 @@ public boolean isIndexed(String field) {
-
     public static final Verifier TEST_VERIFIER = new Verifier(new Metrics());
 
+    private static final Map<String, Map<String, Column>> TABLES;
```
Moved this up since the statics are initialized in declaration order - without this change the TEST_CONFIG always receives a NULL table since the static block isn't yet ran.

costin commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseLexer.g4:
```diff
@@ -344,19 +344,19 @@ ENRICH_FIELD_WS
 
-LOOKUP_QUOTED_IDENTIFIER
-    : QUOTED_IDENTIFIER -> type(QUOTED_IDENTIFIER)
+LOOKUP_FIELD_ID_PATTERN
+    : ID_PATTERN -> type(ID_PATTERN)
```
Used ID_PATTERN since it's likely in the future will use the same syntax with ENRICH (WITH foo*).
The other changes are necessary due to the way lexing works since we want to avoid ambiguities (UNQUOTED IDENTIFIER and ID_PATTERN overlap).

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -119,16 +119,19 @@ public class Analyzer extends ParameterizedRuleExecutor<LogicalPlan, AnalyzerCon
+        var init = new Batch<>(
+            "Initialize",
+            Limiter.ONCE,
             new ResolveTable(),
             new ResolveEnrich(),
-            new ResolveFunctions(),
-            new ResolveRefs(),
-            new ImplicitCasting()
+            new ResolveLookup(),
+            new ResolveFunctions()
         );
+
+        var resolution = new Batch<>("Resolution", new ResolveRefs(), new ImplicitCasting());
```
This is unrelated to this PR - the 'init' rules will resolve things on their first run thus it's a waste to keep using them in a loop.
The same applies for the newly introduced rule for lookup which resolves the local relationship but does NOT perform field resolution - that is left for ResolveRefs.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -316,6 +319,54 @@ private static NamedExpression createEnrichFieldExpression(
+        protected LogicalPlan rule(Lookup lookup, AnalyzerContext context) {
+            // the parser passes the string wrapped in a literal
+            Source source = lookup.source();
+            Expression tableNameExpression = lookup.tableName();
+            String tableName = lookup.tableName().toString();
+            Map<String, Map<String, Column>> tables = context.configuration().tables();
+            LocalRelation localRelation = null;
+
+            if (tables.containsKey(tableName) == false) {
+                String message = "Unknown table [" + tableName + "]";
+                // typos check
+                List<String> potentialMatches = StringUtils.findSimilar(tableName, tables.keySet());
+                if (CollectionUtils.isEmpty(potentialMatches) == false) {
+                    message = UnresolvedAttribute.errorMessage(tableName, potentialMatches).replace("column", "table");
+                }
+                tableNameExpression = new UnresolvedAttribute(tableNameExpression.source(), tableName, null, message);
+            }
```
A trick used in this block, similar to enrich is to pass the user data as a literal and then convert it into a unresolved attribute if it's not found.

costin replied:
Look closer - the block above shows why an Expression is used instead of a Literal.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -316,6 +319,54 @@ private static NamedExpression createEnrichFieldExpression(
+            return new Lookup(source, lookup.child(), tableNameExpression, lookup.matchFields(), localRelation);
+        }
+
+        private LocalRelation tableMapAsRelation(Source source, Map<String, Column> mapTable) {
```
By converting the map into a localrelation, we can use the usual resolution infrastructure.

nik9000 replied:
Should we do this up front to all `table`s when building the config?

costin replied:
Sorry, just saw this message now: this resolution of tables can be done in the PreAnalyze phase, similar to how we do it for enrich or the rest of the indices.

Added a task for it in #109353

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -490,6 +495,39 @@ private LogicalPlan resolveMvExpand(MvExpand p, List<Attribute> childrenOutput)
+            List<NamedExpression> resolvedOn = new ArrayList<>(on.size());
+            List<Attribute> localOutput = l.localRelation().output();
+
+            for (NamedExpression ne : on) {
+                if (ne instanceof UnresolvedAttribute ua && ua.customMessage() == false) {
+                    ne = maybeResolveAttribute(ua, localOutput);
+                    // can't find the field inside the local relation
+                    if (ne instanceof UnresolvedAttribute lua) {
+                        // adjust message
+                        ne = lua.withUnresolvedMessage(
+                            lua.unresolvedMessage().replace("Unknown column", "Unknown column in lookup target")
+                        );
+                    }
+                    // check also the child output by resolving to it
+                    else {
+                        ne = maybeResolveAttribute(ua, localOutput);
+                    }
+                }
+
+                resolvedOn.add(ne);
+            }
```
Just a long winded way of doing resolution for better error messages (missing in the target lookup table vs the child/main table).

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:
```diff
@@ -125,7 +126,21 @@ else if (p.resolved()) {
-            } else {
+            }
+            // similar approach for Lookup
+            else if (p instanceof Lookup lookup) {
+                // first check the table
+                var tableName = lookup.tableName();
+                if (tableName instanceof Unresolvable u) {
+                    failures.add(fail(tableName, u.unresolvedMessage()));
+                }
+                // only after that check the match fields
+                else {
+                    lookup.matchFields().forEach(unresolvedExpressions);
+                }
+            }
```
Added this in to avoid telling the fields are not found if the table does not exist (which is evident).

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/ExpressionBuilder.java:
```diff
@@ -224,8 +226,34 @@ public UnresolvedAttribute visitQualifiedName(EsqlBaseParser.QualifiedNameContex
     @Override
-    public List<Attribute> visitQualifiedNames(EsqlBaseParser.QualifiedNamesContext ctx) {
-        return visitList(this, ctx.qualifiedName(), Attribute.class);
+    public List<NamedExpression> visitQualifiedNamePatterns(EsqlBaseParser.QualifiedNamePatternsContext ctx) {
+        return visitQualifiedNamePatterns(ctx, ne -> {});
+    }
+
+    protected List<NamedExpression> visitQualifiedNamePatterns(
+        EsqlBaseParser.QualifiedNamePatternsContext ctx,
+        Consumer<NamedExpression> checker
+    ) {
+        if (ctx == null) {
+            return emptyList();
+        }
+        List<EsqlBaseParser.QualifiedNamePatternContext> identifiers = ctx.qualifiedNamePattern();
+        List<NamedExpression> names = new ArrayList<>(identifiers.size());
+
+        for (EsqlBaseParser.QualifiedNamePatternContext patternContext : identifiers) {
+            names.add(visitQualifiedNamePattern(patternContext, checker));
+        }
+
+        return names;
+    }
+
+    protected NamedExpression visitQualifiedNamePattern(
+        EsqlBaseParser.QualifiedNamePatternContext patternContext,
+        Consumer<NamedExpression> checker
+    ) {
+        NamedExpression ne = visitQualifiedNamePattern(patternContext);
+        checker.accept(ne);
```
Massaged this code a bit since it's used in multiple places and only the validation error differs - probably could be simplified even further but that's for another day.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/ExpressionBuilder.java:
```diff
@@ -587,11 +615,12 @@ public NamedExpression visitEnrichWithClause(EsqlBaseParser.EnrichWithClauseCont
+        return visitQualifiedNamePattern(ctx, ne -> {
+            if (ne instanceof UnresolvedNamePattern up) {
+                var src = ne.source();
+                throw new ParsingException(src, "Using wildcards [*] in ENRICH WITH projections is not allowed [{}]", up.pattern());
```
Unrelated to this PR - just use [] everywhere.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Lookup.java:
```diff
@@ -22,86 +23,62 @@
-        if (this.output == null) {
-            this.output = mergeOutputAttributes(mergeValues, child().output());
+        if (lazyOutput == null) {
+            List<? extends NamedExpression> rightSide = localRelation != null ? localRelation.output() : matchFields;
+            lazyOutput = mergeOutputAttributes(child().output(), rightSide);
```
The local relation is not always available (no such table) - thus return the matching attribute instead of some fake/empty output list for the relationship.

nik9000 replied:
This'll fail late on, right?

nik9000 replied:
Which is kind of the point. We want these to generate "unresolved" errors.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/Join.java:
```diff
@@ -0,0 +1,138 @@
+import static org.elasticsearch.xpack.esql.plan.logical.join.JoinTypes.CoreJoinType.LEFT;
+import static org.elasticsearch.xpack.esql.plan.logical.join.JoinTypes.CoreJoinType.RIGHT;
+
+public class Join extends BinaryPlan {
```
The join class from the inlinestats branch.

nik9000 replied:
:+1:

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/JoinType.java:
Same as above

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/JoinTypes.java:
same

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/Mapper.java:
```diff
@@ -161,13 +162,9 @@ private PhysicalPlan map(UnaryPlan p, PhysicalPlan child) {
-                lookup.mergeValues()
+            return new HashJoinExec(lookup.source(), child, null, emptyList(), null, null
+            // lookup.matchValues(),
+            // lookup.mergeValues()
```
This doesn't match anymore since there's no more Lookup - just join and local relationship.
This means the hashjoin can be broadcasted (as data strategy).

costin commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -4681,6 +4681,28 @@ public void testIsNullDisjunction() throws Exception {
+     *   \_LocalRelation[[name{f}#14, int{f}#15],[BytesRefVectorBlock[vector=BytesRefArrayVector[positions=10]], IntVectorBlock[vector=I
+     * ntArrayVector[positions=10, values=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]]]]]
+     */
+    public void testBasicLookup() throws Exception {
```
How the plan look like with this commit.

costin commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/StatementParserTests.java:
Unrelated - picking up the ( to [ changes. Should have kept this as a separate PR but didn't as it seems the overhead would be too big.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/Column.java:
```diff
@@ -0,0 +1,39 @@
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql;
```
I thought about it but it'd be alone! I don't want the class to be lonely.

In all honestly I figured we'd move it when we came up with more table-like-things.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/JoinTypes.java:
```diff
@@ -0,0 +1,95 @@
+     * each side of the join are equal).
+     * One important difference is that the USING clause does not include the join columns in the result set.
+     */
+    public static class UsingJoinType implements JoinType {
```
It does feel weird to have USING be something different from a join. Like, maybe it's just

```
record JoinConfig(List<Attribute> mergedColumns, List<Expression> extraConditions, JoinType left/right/outer/whatever)
```

And right now we don't support any `extraConditions`. And `mergedColumns` has to be non-empty unless we're doing the FULL join on a single row thing.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/JoinTypes.java:
```diff
@@ -0,0 +1,95 @@
+        LEFT("LEFT OUTER"),
+        RIGHT("RIGHT OUTER"),
+        FULL("FULL OUTER"),
+        CROSS("CROSS");
```
I'm a little uncomfortable adding all of these enum constants when we don't have any way to run them. But, like, these are the kinds of joins in SQL and they'd been like that for a while. So it's probably fine.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Lookup.java:
```diff
@@ -22,86 +23,62 @@
+    private final List<NamedExpression> matchFields;
+    // initialized during the analysis phase for output and validation
+    // afterward, it is converted into a Join (BinaryPlan) hence why here it is not a child
+    private final LocalRelation localRelation;
```
I'd prefer this be a javadoc. And maybe annotate it with `@Nullable`.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Lookup.java:
```diff
@@ -22,86 +23,62 @@
+        this(source, child, tableName, matchFields, null);
+    }
+
+    public Lookup(Source source, LogicalPlan child, Expression tableName, List<NamedExpression> matchFields, LocalRelation localRelation) {
```
The two ctors feels like overkill to me. We can pass `null` in during the tree walking.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/LogicalPlanBuilder.java:
```diff
@@ -411,12 +403,23 @@ private static Tuple<Mode, String> parsePolicyName(Token policyToken) {
+            }
+        }).name(), DataTypes.KEYWORD);
+
+        return p -> new Lookup(source, p, tableName, matchFields);
```
:+1:

That's much nicer.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/LogicalPlanBuilder.java:
```diff
@@ -411,12 +403,23 @@ private static Tuple<Mode, String> parsePolicyName(Token policyToken) {
+        });
+
+        Literal tableName = new Literal(source, visitQualifiedNamePattern(ctx.tableName, ne -> {
+            if (ne instanceof UnresolvedNamePattern || ne instanceof UnresolvedStar) {
```
It's so `instanceof`! I wish there were some named way of saying "precise" or something. That can be a problem for another time though.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/LogicalPlanBuilder.java:
```diff
@@ -327,21 +322,18 @@ public PlanFactory visitRenameCommand(EsqlBaseParser.RenameCommandContext ctx) {
-        boolean hasSeenStar = false;
-        for (QualifiedNamePatternContext patternContext : identifiers) {
-            NamedExpression ne = visitQualifiedNamePattern(patternContext);
+        final Holder<Boolean> hasSeenStar = new Holder<>(false);
```
Oh interesting. In the rare cases I need this I tend to do `boolean[] hasSeenStar = new boolean[] {false};` - but this `Holder` dude is thing SQL long ago. It has a name which is nice.

nik9000 committed:
Merge branch 'main' into lookup_real___

nik9000 committed:
MErge

nik9000 committed:
precommit

nik9000 committed:
Undo

nik9000 committed:
Merge branch 'main' into lookup_real___

nik9000 committed:
Linking

nik9000 committed:
More

nik9000 committed:
Merge branch 'main' into lookup_real___

nik9000 committed:
FLow

nik9000 committed:
CSV passing

nik9000 committed:
More

nik9000 committed:
More tests

nik9000 committed:
More tests

nik9000 commented:
Costin pushed a change to this branch to model the `LOOKUP` as a full blown `Join` that we then identify as something we can implement as a `HashJoinExec`. I've pushed a few PRs that smoothed out a bunch of fun edges.

The interesting thing is that `Join` is our first `BinaryOperator` - it has two child operators - the rest of the ESQL query on one side and the `table` provided in the parameter on the other.  Everything else in ESQL is `Unary` - the `Driver` itself is fundamentally a unary entity at the moment.

We solve this by making `HashJoinExec` unary - just like every other hash join ever implemented it prepares the hash table up front and then streams values through it. That "preparation" is just something we do at construction time. That's kind of a hack, but it works and it let's us build things. We can replace that with some kind of actual binary-operation `Driver` later.

I haven't resolved serialization - it's important that the `LocalRelation`s that this uses only be serialized to the data nodes:
1. If they are used
2. One time

Right now I just blow up if you try to serialize this. Which is a shame because one of the nice parts of `HashJoinExec` is that you can execute it on the data node or coordinating node. That works, but I have to fix serialization.

There are lots of optimizations that don't work super well and will need some followup work. In particular, it looks like we can't "project" "through" a `Join` and down into the plans below it. That's solveable but we don't have to do it to get this prototype in.

nik9000 committed:
Merge branch 'main' into lookup_real___

nik9000 committed:
precommit almost

nik9000 committed:
WIP

nik9000 committed:
Revert "WIP"

This reverts commit 80602a28594ba334ee452cec2d5bcce4578a3a15.

nik9000 committed:
WIP

nik9000 committed:
Merge branch 'main' into lookup_real___

nik9000 committed:
WIP

nik9000 committed:
Almost

nik9000 committed:
More tests

nik9000 commented:
Progress report:

Serialization works now.

My biggest open question is this trick I do where I rework `FieldAttributes` from the local relation into `ReferenceAttributes`. This prevents looking up those attributes in `SearchStats`. I *think* the right solution to this is to modify how we use `SearchStats` to lookup information based on where the `FieldAttribute` comes from. Or maybe put a reference to some `Stats` in the `FieldAttribute` itself. But, for now, converting to a `ReferenceAttribute` seems to solve the problem. I'd love to hear more from experts 

I'm updating tests and removing NOCOMMITs today and might finish. Maybe!

nik9000 commented:
Another open question - we puts LOOKUP columns *before* the incoming data. I think maybe we should flip that.

nik9000 committed:
Tests

nik9000 committed:
TEsts

nik9000 committed:
Remove NOCOMMITs

These are in comments in the PR

nik9000 committed:
Printlnbegone

nik9000 committed:
Merge branch 'main' into lookup_real___

nik9000 committed:
WIP

nik9000 committed:
Docs!

nik9000 committed:
Little format

nik9000 committed:
No, I like this format

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

nik9000 committed:
Merge branch 'main' into lookup_real___

nik9000 committed:
Merge branch 'main' into lookup_real___

nik9000 committed:
spotless

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/topN.csv-spec:
```diff
@@ -134,3 +134,37 @@ Otmar             |Herbst           |[-8.19, -1.9, -0.32]        |[-0.32, -1.90,
+| WHERE emp_no < 10005
+| SORT languages ASC, emp_no ASC
+| RENAME languages AS int
+| LOOKUP int_number_names ON int
```
I can't seem to find any csv tests with `| LOOKUP in_number_names ON int WITH ... AS ...`, which I'd expect should work similarly to ENRICH. I think it's important that the columns can be given chosen names in the lookup command, and we should test this.

nik9000 replied:
I haven't implemented the renaming. I'm not 100% sure you need to be able replace because you normally have control over the table itself in the request. Except. I don't think everyone will. So it'll be important. It's important here because of the test making the table so general. But I'd imagine a table being defined by, say, a kibana plugin.

alex-spies replied:
We can do this in a follow-up, as long as we don't bake the assumption of always getting the attribute names from the table columns into the execution somehow.

I think it's important as renaming table columns may not always be possible/ergonomic + grok/dissect currently have a problem where you cannot rename the generated attributes and it causes complications in optimization rules. (The column attributes may shadow e.g. field attributes and then you have to awkwardly rename stuff - in terms of opt. rules that means introducing a projection which you may not want.)

nik9000 replied:
This'll be a followup.

alex-spies reviewed (commented):
Gave that a first go and will do another round of reviewing on Monday.

nik9000 committed:
Update docs/reference/esql/esql-commands.asciidoc

Co-authored-by: Alexander Spies <alexander.spies@elastic.co>

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -653,6 +658,11 @@ protected LogicalPlan rule(Limit limit) {
                         }
                     }
                 }
+            } else if (limit.child() instanceof Join join) {
```
This was actually a suggestion from Costin. I didn't start this way. I do think it'd be useful to do things like push down things to the `LocalRelation`, but this is something I've got to trust the QL experts in.

alex-spies replied:
As discussed, I think a good compromise would be to only implement the specialized `LeftJoin` for now - so the fact that it's a left join is encoded in the class rather than in an external config. That allows to implement methods on it which may not apply to general joins, and we can more easily treat `Lookup` similarly to `Enrich` in optimization rules.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Lookup.java:
```diff
@@ -0,0 +1,134 @@
+
+/**
+ * Looks up values from the associated {@code tables}.
+ * The class is used only during the analysis phased, after which it is replaced by a {@code Join}.
```
nit: Analysis is resolving types and fields - we technically replace in the substitution phase of logical optimization.
```suggestion
 * The class is supposed to be substituted by a {@link Join}.
```

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Lookup.java:
```diff
@@ -0,0 +1,134 @@
+ */
+public class Lookup extends UnaryPlan {
+    private final Expression tableName;
+    private final List<NamedExpression> matchFields;
```
Sorry, I made a mistake. I thought these were the expressions for the looked up columns, but this is what we match/join on. This is fine and follows the same pattern as Enrich.

What I was trying to suggest was having a `List<Alias>` that refers to the table columns and allows to rename the columns. Also, we may need a way to add only a subset of the columns - but both of these things can be done in the future.

(`NamedExpression` is more flexible - in case of `Enrich`, the enrich fields are encoded as such and, in practice, are either a `ReferenceAttribute` (enrich field without `WITH`) or `Alias` (enrich field with `WITH`).)

alex-spies reviewed (commented):
Had a second go at it and left some more opinions. Will continue with at least another go!

alex-spies commented on x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/EsqlSpecTestCase.java:
```diff
@@ -154,19 +158,21 @@ protected void shouldSkipTest(String testName) throws IOException {
-            testCase.expectedWarnings(false),
-            testCase.expectedWarningsRegex()
-        );
+        if (testCase.query.contains("LOOKUP")) {
```
Maybe let's make this case insensitive.

nik9000 replied:
:+1:

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Lookup.java:
```diff
@@ -0,0 +1,134 @@
+        LogicalPlan child,
+        Expression tableName,
+        List<NamedExpression> matchFields,
+        @Nullable LocalRelation localRelation
```
This theoretically allows for lookups with multivalued blocks. I wonder if we want to
- Disable this through validation, or
- somehow test this?

We may not want to allow `LOOKUP` where the supplied table has multivalues, but later `INLINESTATS` will force us to consider this if we want to allow `INLINESTATS` with `VALUES`, no?

alex-spies replied:
I also realized: what happens when the key column is multivalued? And how's it work when the key is composite - how do we define a match - only when all values coincide, or does `a: [1,2], b:[3,4]` in the index match with `a:1, b:4` in the table?

nik9000 replied:
We have a *ton* of tests around multivalued keys and values actually. But, yeah, we should disable multivalued table keys, at least for now. We don't need them for INLINESTATS because that'll always join on the grouping columns which never output multivalued fields.

Now, when the incoming field is multivalued, that gets *fun*. We match the combinatorial explosion of all combinations of the keys. Specifically because that's what INLINESTATS will need - at least, that's what it'll need if we want it to mirror behavior of aggs. Even more fun - there isn't any real ordering in that combinatorial explosion. So, while it's correct, it might not be useful.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanStreamOutput.java:
```diff
@@ -80,7 +80,6 @@ public PlanStreamOutput(
     }
 
     public void writeLogicalPlanNode(LogicalPlan logicalPlan) throws IOException {
-        assert logicalPlan.children().size() <= 1;
```
Super nit: Maybe this can be left in place?
```java
        assert logicalPlan.children().size() <= 2;
```
or
```java
        assert (logicalPlan.children().size() <= 1) || (logicalPlan instanceOf Join && logicalPlan.children.size() == 2);
```

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/JoinConfig.java:
```diff
@@ -0,0 +1,30 @@
+import java.io.IOException;
+import java.util.List;
+
+public record JoinConfig(JoinType type, List<NamedExpression> unionFields) implements Writeable {
```
I think we could align the naming with Enrich and Lookup:
```suggestion
public record JoinConfig(JoinType type, List<NamedExpression> matchFields) implements Writeable {
```
Applies also to HashJoinExec.

In the context of joins, "union" reminds me of, well, relational unions, which are a different concept.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/JoinType.java:
```diff
@@ -0,0 +1,52 @@
+import java.io.IOException;
+
+public enum JoinType implements Writeable {
+    INNER(0, "INNER"),
+    LEFT(1, "LEFT OUTER"),
+    RIGHT(2, "RIGHT OUTER"),
+    FULL(3, "FULL OUTER"),
+    CROSS(4, "CROSS");
```
I don't think we need all the join types, especially as we only implement `LEFT OUTER` for now.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/Join.java:
```diff
@@ -0,0 +1,144 @@
+    private List<Attribute> computeOutput() {
+        List<Attribute> right = makeReference(right().output());
+        return switch (config.type()) {
+            case LEFT -> // right side becomes nullable
+                NamedExpressions.mergeOutputAttributes(left().output(), makeNullable(right));
```
I think this gives precedence to the attributes from `left().output()` in case of name collisions; I think this is the opposite of Enrich's variable shadowing behavior, which is why I think we should follow what Enrich does.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/Join.java:
```diff
@@ -0,0 +1,144 @@
+        return switch (config.type()) {
+            case LEFT -> // right side becomes nullable
+                NamedExpressions.mergeOutputAttributes(left().output(), makeNullable(right));
+            case RIGHT -> // left side becomes nullable
+                NamedExpressions.mergeOutputAttributes(makeNullable(left().output()), right);
+            case FULL -> // both sides become nullable
+                NamedExpressions.mergeOutputAttributes(makeNullable(left().output()), makeNullable(right));
+            default -> // neither side becomes nullable
+                NamedExpressions.mergeOutputAttributes(left().output(), right);
```
This is mixing how variable shadowing works with the kind of join. Since we won't have right and full joins for now, I don't think this is useful code, as we cannot properly test it, and it's more subtle than it looks like.

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/floats.csv-spec:
```diff
@@ -569,3 +569,31 @@ s:double | emp_no:integer | salary:integer  | salary_change:double
+
+double:double | name:keyword
+         2.08 | two point zero eight
+;
```
Another thing to test:
What happens when the input table has multiple columns with the same names? Do we catch this in validation?

nik9000 replied:
> Some column names from the table used in LOOKUP coincide with names in the index and the values differ; I'd expect that the column's attributes shadow the index' attributes.

Right now that's how we find the join columns. So there shouldn't be shadowing - it should be the values from the original index and `null` results for the lookup output.



> In the table, there's duplicate rows that match a given key.

That's supposed to be an error. I'll make sure of it.

> * Lookups with composite keys: `LOOKUP ... ON a, b, c`. `Lookup.java` has a `List<NamedExpression>` for the `matchFields`, so we should either test multiple match fields or simplify this to just a singl

Yeah. I never wrote that one. The PR got too big. Hard to get good test coverage in those cases.



> What happens when the input table has multiple columns with the same names? Do we catch this in validation?

That's invalid json the way the API is now. I'm not sure the tests even let you emit such json. I can try it though.

nik9000 replied:
Tested in "duplicate column names in table"

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Lookup.java:
```diff
@@ -0,0 +1,134 @@
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(super.hashCode(), tableName, matchFields, localRelation);
```
Thought: Hashing the local relation every time could be expensive, no?

Since the local relation is uniquely identified by the table name, we _could_ consider the `localRelation` to just be part of `Lookup`'s state (resolved vs. unresolved table) - and instead hash the boolean `localRelation == null`, no?
Similarly for `equals`.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/Join.java:
```diff
@@ -0,0 +1,144 @@
+
+public class Join extends BinaryPlan {
+
+    private final JoinConfig config;
```
I think it'd be better if the match fields were part of this `LogicalPlan`'s `expressions`. The way I've seen this modeled in the past is either having a list of pairs of expressions, or two lists of expressions, or more precisely reference attributes. The left (right) ones should refer to attributes of the left (right) plan.

This would mean less special casing for joins e.g. in the dependency checker - the referenced attributes would just be returned by `Join.references()`.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/Join.java:
```diff
@@ -0,0 +1,144 @@
+
+    @Override
+    public boolean expressionsResolved() {
+        return true;
```
This doesn't take into account the validity of the join config. If I'm right, it's possible to define a broken `Join`, referring to non-existing fields in the join config, but have `resolved()` come out without any problems.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/LocalExecutionPlanner.java:
```diff
@@ -480,6 +486,52 @@ private PhysicalOperation planEnrich(EnrichExec enrich, LocalExecutionPlannerCon
+            if (join.child().outputSet().contains(f)) {
+                continue;
+            }
+            layoutBuilder.append(f);
```
Thought: Interestingly, this operator seems to only add columns/channels/attributes, even in case of name collisions. If that's correct, there's opportunity to reduce memory usage by projecting out shadowed attributes.

This applies similarly to e.g. `planGrok`, which also only ever adds new attributes to the layout, even in case of shadowing.

nik9000 replied:
I'm skipping this one for the moment. Maybe a follow up.

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -4681,6 +4684,104 @@ public void testIsNullDisjunction() throws Exception {
+            | RENAME languages AS int
+            | LOOKUP int_number_names ON int
+            """);
+        var join = as(plan, Join.class);
```
We should also make assertions about the join config.
Also applies to `testLookupStats`.

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -4681,6 +4684,104 @@ public void testIsNullDisjunction() throws Exception {
+
+    /**
+     * Expects
+     * <pre>{@code
+     * Limit[1000[INTEGER]]
+     * \_Aggregate[[name{f}#20],[MIN(emp_no{f}#10) AS MIN(emp_no), name{f}#20]]
+     *   \_Join[JoinConfig[type=LEFT OUTER, unionFields=[int{r}#4]]]
+     *     |_EsqlProject[[_meta_field{f}#16, ...]]
+     *     | \_EsRelation[test][_meta_field{f}#16, emp_no{f}#10, first_name{f}#11, ..]
+     *     \_LocalRelation[[name{f}#20, int{f}#21],[BytesRefVectorBlock[vector=BytesRefArrayVector[positions=10]], IntVectorBlock[vector=I
+     * ntArrayVector[positions=10, values=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]]]]]
+     * }</pre>
```
We are not generating javadoc from these, so let's remove the `<pre>` tags to be consistent inside this file.
```suggestion
     * {@code
     * Limit[1000[INTEGER]]
     * \_Aggregate[[name{f}#20],[MIN(emp_no{f}#10) AS MIN(emp_no), name{f}#20]]
     *   \_Join[JoinConfig[type=LEFT OUTER, unionFields=[int{r}#4]]]
     *     |_EsqlProject[[_meta_field{f}#16, ...]]
     *     | \_EsRelation[test][_meta_field{f}#16, emp_no{f}#10, first_name{f}#11, ..]
     *     \_LocalRelation[[name{f}#20, int{f}#21],[BytesRefVectorBlock[vector=BytesRefArrayVector[positions=10]], IntVectorBlock[vector=I
     * ntArrayVector[positions=10, values=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]]]]]
     * }
```
Same for `PhysicalPlanOptimizerTests.java`

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizerTests.java:
```diff
@@ -4138,6 +4150,122 @@ public void testMaxQueryDepthPlusExpressionDepth() {
+        var join = as(outerProject.child(), HashJoinExec.class);
+        assertMap(join.unionFields().stream().map(Object::toString).toList(), matchesList().item(startsWith("int{r}")));
+        assertMap(
+            join.output().stream().map(Object::toString).filter(s -> s.contains("int") || s.contains("name")).sorted().toList(),
```
Why do we filter the output list here?

alex-spies reviewed (changes requested):
Thanks @nik9000 , this is great.

Before we merge this, I think we should take care of the following:
- IMHO the `Join` logical plan class needs more love. I think it should treat match fields as references, not as a mere config. Also, we should reduce to the actual join types we have; in contrast to SQL, the interaction with attribute shadowing is subtle, so trying to anticipate the typical relation join types (left, right, outer, cross, inner) is IMHO premature and will really make writing optimization rules for `LOOKUP` more complicated than necessary. IMHO `Join` is not even necessary, the `Lookup` logical plan could also be mapped to `HashJoinExec` directly, but at least specializing to `LeftJoin` is fine.
- Analogously to `ENRICH`, the attributes generated by `LOOKUP` should be on the right hand side; ESQL has so far been quite consistent with its left-to-right semantics, where new attributes are added on the right hand side and shadow attributes left of them.
- We should include more edge cases, esp. in our csv tests; we can also add them as awaiting fixes, for now, and implement the correct edge case handling in follow-ups.

nik9000 commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/floats.csv-spec:
```diff
@@ -569,3 +569,31 @@ s:double | emp_no:integer | salary:integer  | salary_change:double
 -1.0     | 10004          | 36174           | [-0.35, 1.13, 3.65, 13.48]
 -1.0     | 10005          | 63528           | [-2.14, 13.07]
 ;
+
```
Probably. I figured it was cross cutting to the types. It is, but it probably still is better to put them in lookup.csv-spec.

nik9000 commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/floats.csv-spec:
```diff
@@ -569,3 +569,31 @@ s:double | emp_no:integer | salary:integer  | salary_change:double
+| LOOKUP double_number_names ON name
+;
+
+double:double | name:keyword
+         2.08 | two point zero eight
```
Yeah. That's one of the things I called out as a known issue with how I've got it. It's do with the way attributes are merged. I tried untangling it but it made everything worse. I think I'll have another go at it soon.

nik9000 commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/floats.csv-spec:
```diff
@@ -569,3 +569,31 @@ s:double | emp_no:integer | salary:integer  | salary_change:double
+FROM employees
+| SORT emp_no
+| LIMIT 4
+| RENAME height AS double
+| LOOKUP double_number_names ON double
```
Maybe, yeah. I think we're only dealing with `=`, at least for now.

nik9000 commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseLexer.g4:
```diff
@@ -339,6 +340,33 @@ ENRICH_FIELD_WS
     : WS -> channel(HIDDEN)
     ;
 
+// LOOKUP ON key
+mode LOOKUP_MODE;
```
I have no idea. Costin seemed to like it this way so I did it. I think LOOKUP is different, but I'm not familiar enough with `mode` in ANTLR to know anything about if this is a good idea.

nik9000 replied:
@costin will have to comment on this one. I don't know that its strictly needed but I don't know if its expensive either.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/Mapper.java:
```diff
@@ -113,6 +117,24 @@ public PhysicalPlan map(LogicalPlan p) {
+            var left = map(bp.left());
+            var right = map(bp.right());
+
+            if (left instanceof FragmentExec) {
+                if (right instanceof FragmentExec) {
+                    throw new EsqlIllegalArgumentException("can't plan binary [" + p.nodeName() + "]");
+                }
+                // in case of a fragment, push to it any current streaming operator
+                return new FragmentExec(p);
+            }
+            if (right instanceof FragmentExec) {
+                // in case of a fragment, push to it any current streaming operator
+                return new FragmentExec(p);
+            }
```
Yeah, I mirrored the other code and it led to quite the abstraction layers. 

To be honest this is all stuff I'm a lot less familiar with.

costin replied:
> we won't have general binary plans, because we cannot map them.

We don't have _yet_ binary plans, but this PR is the first step in the right direction. see https://github.com/elastic/elasticsearch/issues/99841

alex-spies reviewed (commented):
One more thing I noticed: we don't seem to validate that datatypes match between the `ON` column and the corresponding column from the table:
```
curl -u elastic:password -H "Content-Type: application/json" "127.0.0.1:9200/_query?format=txt" -d '{                                                                    9:07
    "query": "ROW a=1::KEYWORD | LOOKUP t ON a",        
    "tables": {
        "t": {
            "a:long":     [    1,     4,     2],
            "v1:integer": [   10,    11,    12],
            "v2:keyword": ["cat", "dog", "wow"]
        }
    },
    "version": "2024.04.01"
}'
```
That just fails the lookup, producing nulls instead.
```
      v1       |      v2       |       a       
---------------+---------------+---------------
null           |null           |1
```

nik9000 committed:
Merge branch 'main' into lookup_real___

nik9000 committed:
spotless

nik9000 committed:
Tests

nik9000 committed:
Merge remote-tracking branch 'nik9000/lookup_real___' into lookup_real___

nik9000 committed:
Update x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Lookup.java

Co-authored-by: Alexander Spies <alexander.spies@elastic.co>

nik9000 committed:
more tests

nik9000 committed:
More

nik9000 committed:
Merge branch 'main' into lookup_real___

nik9000 committed:
Compile plz

nik9000 commented:
> One more thing I noticed: we don't seem to validate that datatypes match between the `ON` column and the corresponding column from the table:

Oh neat. Yeah, that's bad.

nik9000 committed:
Multivalued keys

nik9000 committed:
More tests

costin commented:
Had a discussion with Alex - we'll revisit the QL side once this PR lands in.
There are several aspects that could be improved such as:
- keeping the join fields as an expression (even if it's just Equals) instead of separate fields since this helps the planner
- adding the missing USING back in the join type (seems it was removed)
- look at the shadowing mechanism in joins

nik9000 committed:
Merge branch 'main' into lookup_real___

nik9000 commented on x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/EsqlSpecTestCase.java:
```diff
@@ -323,4 +329,56 @@ public static void assertRequestBreakerEmpty() throws Exception {
+            na.appendInt(100);
+            na.appendInt(-100);
+
+         */
```
Leftover.

nik9000 committed:
Merge branch 'main' into lookup_real___

nik9000 committed:
Merge branch 'main' into lookup_real___

nik9000 commented:
OK! I was distracted last week. But this week I'm going to try to land this one. So! First thing is to merge main into it and get the tests passing. Next I need to corral all of the line comments into a big check list. There are a lot of line comments we've said "nope, we're not doing this" and there a lot we want to do and I've lost track. So I need to make a list.

nik9000 committed:
Finish merge

elasticsearchmachine commented:
Pinging @elastic/kibana-esql (ES|QL-ui)

nik9000 committed:
Merge branch 'main' into lookup_real___

nik9000 committed:
Drop

nik9000 commented:
TODOS:
* [ ] https://github.com/elastic/elasticsearch/pull/107987#issuecomment-2118394344
* [ ] https://github.com/elastic/elasticsearch/pull/107987#discussion_r1587059151
* [ ] https://github.com/elastic/elasticsearch/pull/107987#discussion_r1601562300
* [ ] https://github.com/elastic/elasticsearch/pull/107987#discussion_r1601583844
* [ ] https://github.com/elastic/elasticsearch/pull/107987#discussion_r1601564991
* [ ] https://github.com/elastic/elasticsearch/pull/107987#discussion_r1601539235
* [ ] https://github.com/elastic/elasticsearch/pull/107987#discussion_r1601531116
* [ ] https://github.com/elastic/elasticsearch/pull/107987#discussion_r1601491776
* [ ] https://github.com/elastic/elasticsearch/pull/107987#discussion_r1600263894
* [ ] https://github.com/elastic/elasticsearch/pull/107987#discussion_r1600260873
* [ ] https://github.com/elastic/elasticsearch/pull/107987#discussion_r1600255564
* [ ] https://github.com/elastic/elasticsearch/pull/107987#discussion_r1600137357
* [ ] https://github.com/elastic/elasticsearch/pull/107987#discussion_r1601510974
* [ ] https://github.com/elastic/elasticsearch/pull/107987#discussion_r1600218810
* [ ] https://github.com/elastic/elasticsearch/pull/107987#discussion_r1600198174
* [ ] https://github.com/elastic/elasticsearch/pull/107987#discussion_r1600166496

nik9000 committed:
Compile pleaze

nik9000 committed:
Apply suggestions from code review

Co-authored-by: Alexander Spies <alexander.spies@elastic.co>

nik9000 committed:
Fix after renames

nik9000 committed:
This breaks it

nik9000 committed:
Fixed

nik9000 committed:
Move results to the right

nik9000 committed:
More tests

nik9000 committed:
Drop filters in tests

nik9000 committed:
Check conditions resolved

nik9000 committed:
Fix test and format

nik9000 committed:
More tests

nik9000 committed:
Check types

nik9000 committed:
Spotless

nik9000 commented:
I'd like to hold https://github.com/elastic/elasticsearch/pull/107987#discussion_r1601539235 until a followup, I think. Otherwise, I believe this is ready for review. I'll need to double check it myself, but I believe I've knocked out the comments I could find.

nik9000 commented:
It occurred to me this morning that I'm going to need a new capability and probably a transport version. That'll be a today thing. And finding tests for it.

nik9000 committed:
Merge branch 'main' into lookup_real___

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/Join.java:
```diff
@@ -0,0 +1,167 @@
+        return switch (config.type()) {
+            case LEFT -> // right side becomes nullable
+                mergeOutput(left().output(), makeNullable(right));
+            case RIGHT -> // left side becomes nullable
+                mergeOutput(makeNullable(left().output()), right);
+            case FULL -> // both sides become nullable
+                mergeOutput(makeNullable(left().output()), makeNullable(right));
+            default -> // neither side becomes nullable
+                mergeOutput(left().output(), right);
```
Can we make the other cases throw an IAE? This way, at least this stays reasonably easy to verify and we don't add dead code.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/Join.java:
```diff
@@ -0,0 +1,167 @@
+    }
+
+    public boolean duplicatesResolved() {
+        return left().outputSet().intersect(right().outputSet()).isEmpty();
```
Just wanted to point out: I think this uses name ids to compute the intersection, not names; so I expect this to return `true` essentially always.

nik9000 replied:
I'll have to go rereview this one.

nik9000 replied:
This one always seems to return `true`. They never intersect so far as I can tell.

costin replied:
Then let's remove it.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Lookup.java:
```diff
@@ -0,0 +1,155 @@
+
+    public JoinConfig joinConfig() {
+        List<Expression> conditions = new ArrayList<>(matchFields.size());
+        List<Attribute> rhsOutput = Join.makeReference(localRelation.output());
```
This is indeed weird. Can't we just keep the field attributes as-is?

Reference attributes normally are needed when we create new attributes, but these are _not_ coming from fields - like in `| eval x = ...`, here `x = ...` is modeled via an `Alias` and calling `xAlias.toAttribute()` or `Expressions.asAttributes(xAlias)` will provide a corresponding `ReferenceAttribute`.

alex-spies replied:
Update: you made a nice comment in tests that explains why these need to be reference attributes, so never mind.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Lookup.java:
```diff
@@ -0,0 +1,155 @@
+            List<? extends NamedExpression> rightSide = localRelation != null
+                ? Join.makeNullable(Join.makeReference(localRelation.output()))
+                : matchFields;
+            lazyOutput = mergeOutputAttributes(child().output(), rightSide);
```
Enrich has the arguments here exactly the other way around, and I think this contributes to the buggy shadowing.

alex-spies reviewed (commented):
Need to do a second pass; found something that requires a follow-up, but shouldn't block this.

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/AnalyzerTests.java:
```diff
@@ -1879,16 +1885,70 @@ public void testInOnText() {
+        var limit = as(plan, Limit.class);
+        assertThat(limit.limit().fold(), equalTo(1000));
+
+        var lookup = as(limit.child(), Lookup.class);
```
Can we also assert `lookup.output()`?

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/lookup.csv-spec:
```diff
@@ -1,5 +1,117 @@
+
+aa:keyword | ab:keyword | na:integer | nb:integer
+foo        | zoi        |       null |       null
+;
```
Maybe it's worth adding a test that does two lookups, maybe even using a looked up column as match field.

nik9000 replied:
:+1:

nik9000 replied:
This is actually broken. I've added it to meta issue.

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -4973,6 +4977,126 @@ public void testIsNullDisjunction() throws Exception {
+     * Expects
+     * <pre>{@code
+     * Limit[1000[INTEGER]]
+     * \_Join[JoinConfig[type=LEFT OUTER, unionFields=[int{r}#4]]]
```
This needs updating (similarly below).

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -4973,6 +4977,126 @@ public void testIsNullDisjunction() throws Exception {
+        assertThat(join.config().type(), equalTo(JoinType.LEFT));
+        assertThat(join.config().matchFields().stream().map(Object::toString).toList(), matchesList().item(startsWith("int{r}")));
+        assertThat(join.config().conditions().size(), equalTo(1));
+        Equals eq = (Equals) join.config().conditions().get(0);
```
super nit: We normally use `as` in tests:
```suggestion
        Equals eq = as(join.config().conditions().get(0), Equals.class);
```

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -4973,6 +4977,126 @@ public void testIsNullDisjunction() throws Exception {
+        assertThat(join.config().matchFields().stream().map(Object::toString).toList(), matchesList().item(startsWith("int{r}")));
+        assertThat(join.config().conditions().size(), equalTo(1));
+        Equals eq = (Equals) join.config().conditions().get(0);
+        assertThat(eq.left().toString(), startsWith("int{r}"));
+        assertThat(eq.right().toString(), startsWith("int{r}"));
```
nit: we could assert that the left `int` actually `semanticEquals` a field from the left input (and analogously for the right hand side).

`int{r}` should have a correct name id, otherwise we might point to the wrong input attribute. (I know this is no fun and quite complex :/ )

nik9000 replied:
I can try that, yeah. I didn't know about `semanticEquals`, better than the toString.

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -4973,6 +4977,126 @@ public void testIsNullDisjunction() throws Exception {
+                /*
+                 * It's important that name is returned as a *reference* here
+                 * instead of a field. If it were a field we'd use SearchStats
+                 * on it and discover that it doesn't exist in the index. It doesn't!
+                 * We don't expect it to. It exists only in the lookup table.
```
Aha! That comment would be immensely helpful as part of the other comment on `Join.makeReference`.

nik9000 replied:
I'll add it there!

alex-spies reviewed (approved):
Except for some minor comments, I think this should be merged and iterated on once it's in main.
Main things to follow up are the var shadowing and adapting optimizer rules for enrich to lookup.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/physical/HashJoinExec.java:
```diff
@@ -0,0 +1,137 @@
+        return joinData;
+    }
+
+    public List<NamedExpression> unionFields() {
+        return matchFields;
```
nit:
```suggestion
    public List<NamedExpression> matchFields() {
        return matchFields;
```

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizerTests.java:
```diff
@@ -4080,6 +4093,145 @@ public void testMaxQueryDepthPlusExpressionDepth() {
+        assertMap(
+            join.output().stream().map(Object::toString).toList(),
+            matchesList().item(startsWith("_meta_field{f}"))
+                .item(startsWith("emp_no{f}"))
+                .item(startsWith("first_name{f}"))
+                .item(startsWith("gender{f}"))
+                .item(startsWith("job{f}"))
+                .item(startsWith("job.raw{f}"))
+                .item(startsWith("int{r}"))
+                .item(startsWith("last_name{f}"))
+                .item(startsWith("long_noidx{f}"))
+                .item(startsWith("salary{f}"))
+                .item(startsWith("name{r}"))
```
This test indicates that we can update `LogicalPlanOptimizer.PruneColumns` - the operator is carrying over way too many columns that are not needed from the index.

Another thing that we need to get right is `PropagateEmptyRelation`: when all of the columns are projected away, we shouldn't even execute the hashjoin.

Similarly, when we don't need any of looked up columns, the `Lookup` can be optimized away altogether.

These optimizations are far from critical now (getting push downs right is!), but should be revisited when we want to remove the `experimental` label from `LOOKUP`.

Should we put this into a meta issue?

nik9000 replied:
https://github.com/elastic/elasticsearch/issues/109353

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizerTests.java:
```diff
@@ -4080,6 +4093,145 @@ public void testMaxQueryDepthPlusExpressionDepth() {
+        ExchangeExec exchange = as(outerTopN.child(), ExchangeExec.class);
+        FragmentExec frag = as(exchange.child(), FragmentExec.class);
+
+        LogicalPlan opt = logicalOptimizer.optimize(frag.fragment());
```
Watch out: this doesn't match reality.

In production, we use a `LocalLogicalPlanOptimizer` to optimize the fragment, so we're not really testing the right thing here (different optimization rules).
We could take inspiration from the csv tests for how to make this test closer to Prod.

alex-spies commented on x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/150_lookup.yml:
```diff
@@ -1,32 +1,245 @@
+          - { "data": 1, "count": 40, "data_d": 1, "count_d": 40, "time": 1674835275223, "color": "red" }
+          - { "index": { } }
+          - { "data": 2, "count": 42, "data_d": 2, "count_d": 42, "time": 1674835275224, "color": "blue" }
+          - { "index": { } }
+          - { "data": 1, "count": 44, "data_d": 1, "count_d": 44, "time": 1674835275225, "color": "green" }
+          - { "index": { } }
+          - { "data": 2, "count": 46, "data_d": 2, "count_d": 46, "time": 1674835275226, "color": "red" }
```
Do we need all the other rows, given that the test only uses the ~~last~~ first two?

nik9000 replied:
I can drop them just so it's simpler to read.

nik9000 commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/lookup.csv-spec:
```diff
@@ -1,5 +1,117 @@
+
+keywordByMvInt
+required_capability: lookup
+ROW int=[1, 2, 3]
```
nice

nik9000 replied:
I haven't done this one yet. I'll get to it in the morning.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/JoinConfig.java:
```diff
@@ -0,0 +1,46 @@
+ * @param matchFields fields that are merged from the left and right relations
+ * @param conditions when these conditions are true the rows are joined
+ */
+public record JoinConfig(JoinType type, List<NamedExpression> matchFields, List<Expression> conditions) implements Writeable {
```
I'm not 100% sure having `matchFields and `conditions` is redundant. But that's to do mostly with this being a bigger abstraction then we need. Like, it's obviously not redundant if you have `List.of(new GreaterThan(lhs, rhs))` because we don't want to remove any rows. In `Lookup` it's always redundant though.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -1259,6 +1268,19 @@ protected LogicalPlan rule(Limit plan) {
+        }
+    }
+
     /**
```
We're moving this into the meta issue here: https://github.com/elastic/elasticsearch/issues/109353

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/Join.java:
```diff
@@ -0,0 +1,167 @@
+        };
+    }
+
+    /**
+     * Merge output fields, left hand side wins in name conflicts <strong>except</strong>
+     * for fields defined in {@link JoinConfig#matchFields()}.
+     */
```
Moving this to the followup issue.

nik9000 committed:
From review

nik9000 committed:
Last one?

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/lookup.csv-spec:
```diff
@@ -1,5 +1,117 @@
+
+intIntByKeywordKeyword
+required_capability: lookup
+ROW aa="foo", ab="zoo"
```
Please add a test where aa and ab (as lookup keys) have the same value - `row aa = "foo", ab="foo" | lookup big on aa, ab".
Followed by variations:
 - lookup on field with empty value
 - lookup on function over field (should give validation error) - can be a follow-up.

nik9000 replied:
"lookup on function over field:" - you mean like `LOOKUP foo ON CONCAT(a, b)`?

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -315,10 +323,60 @@ private static NamedExpression createEnrichFieldExpression(
+                String name = entry.getKey();
+                Column column = entry.getValue();
+                // create a fake ES field - alternative is to use a ReferenceAttribute
+                EsField field = new EsField(name, column.type(), Map.of(), false, false);
+                attributes.add(new FieldAttribute(source, null, name, field));
```
The source of the attribute should be decoupled from the Attribute itself. `MetadataAttribute` itself is stretching the concept since it's just a special FieldAttribute (who's name starts with _).

nik9000 replied:
I'm moving this to the followups meta issue because I'm not sure from the words in these two comments if we're happy about this code. I'm not super happy that we have to do this `toReference` dance and think the right way to fix it is to do something about the way we do `SearchStats`. I think that's related here. I think.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/JoinType.java:
```diff
@@ -0,0 +1,52 @@
+            case 2 -> RIGHT;
+            case 4 -> FULL;
+            case 5 -> CROSS;
+            default -> throw new IllegalArgumentException("unsupported join [" + id + "]");
```
We have EsqlIllegalArgumentException to differentiate between exceptions thrown by ESQL code vs other packages/JDK.

costin reviewed (approved):
LGTM - thanks Nik for making this happen and incorporating the various feedback received.

costin commented:
cc @astefan and @bpintea for awareness

nik9000 committed:
Merge branch 'main' into lookup_real___

nik9000 committed:
On function

nik9000 committed:
WIP

nik9000 committed:
don't dupe

nik9000 committed:
more

nik9000 commented:
OK! I've pushed the last update from a comment. I think we've extracted all of the unfinished comments into our lovely followup meta issue (https://github.com/elastic/elasticsearch/issues/109353). I'm marking this `auto-merge` and beginning to land it. Thanks so much friends. 186 comments, 96 commits. Over a little more than a month. And 16 pre-PRs that had to land before that. And followups. This is a big one!

nik9000 committed:
Docs

nik9000 committed:
Merge branch 'main' into lookup_real___

nik9000 committed:
one more test

nik9000 committed:
Merge branch 'main' into lookup_real___

nik9000 committed:
fix mergre

nik9000 committed:
Merge branch 'main' into lookup_real___

nik9000 committed:
Merge branch 'main' into lookup_real___

nik9000 committed:
Merge branch 'main' into lookup_real___

nik9000 committed:
Fix tests

nik9000 commented:
Oh darn. I thought the bwc failures were because of a release, but that doesn't make sense - we didn't have a branch cut. they were real. I fixed them and pushed.
